torch/_higher_order_ops/triton_kernel_wrap.py

```diff
@@ -239,25 +248,49 @@ def convert(token):
     # In alternative representation, function names are quoted.
     # It should be possible to move this into the grammar alltogether.
     def convert_name(token):
+        if token is None:
+            return None
         s = token.value
         if len(s) > 2 and s[0] == '"' and s[-1] == '"':
             return s[1:-1]
         return s
 
     @v_args(inline=True)
     class CalculateOps(Transformer):
-        def process_op_no_ret(self, *args):
-            self.process_op(None, *args)
+        def process_op(self, ret, op_name, fn_name, args, *rest):
+            ops[convert(ret)] = Op(
+                convert_name(op_name), convert_name(fn_name), convert(args)
+            )
 
-        def process_op(self, ret, name, args, *rest):
-            ops[convert(ret)] = Op(convert_name(name), convert(args))
+        def process_func(self, name, *rest):
+            nonlocal ops
+            functions[name.value] = ops
+            ops = {}
 
-    parser = Lark(grammar, parser="lalr", transformer=CalculateOps())
+    parser = Lark(
+        grammar, parser="lalr", maybe_placeholders=True, transformer=CalculateOps()
+    )
     parser.parse(ttir)
-    return ops
+    return functions
 
 
-def analyze_kernel_mutations(ops, kwargs, tensor_param_locs):
+class MemoizeWithCycleCheck:
+    def __init__(self, fn):
+        self.cache = {}
+        self.fn = fn
+
+    def __call__(self, functions, fn_name, num_args):
+        key = (fn_name, num_args)
+        if key not in self.cache:
+            self.cache[key] = None
+            self.cache[key] = self.fn(functions, fn_name, num_args)
+        if self.cache[key] is None:
+            raise Exception("Recursion is not supported")
```

Comment:
Wondering how this can be reached if `key` is either already in `self.cache` or we set it two lines above? `self.fn` can return `None`?

Reply:
explained above

Reply:
Should we bail out instead of raising an exception if the user-written Triton kernel has a recursion in it (that is if the Triton kernel code is otherwise valid)? Or does Triton not support recursion and here we're just honoring that?

Reply:
Raising exception will result in bailing out since the parent function has a try-catch. I do this in other places of this analysis too. Exception based control flow is very pythonic.

Triton probably supports exception but it would be harder to deal with some edge cases, so I choose to bail out.